### PR TITLE
Initial idea for disabling ranges

### PR DESCRIPTION
**Pull request by davidtheclark**
NOT INTENDED TO BE MERGED YET. I'm posting this so it can be reviewed as it is so far, more commits added, etc.

This is a start for the feature discussed in https://github.com/stylelint/stylelint/issues/105 --- disabling rules for sections of code between special comments.

Notice that I added a function `report()` that wraps PostCSS's `result.warn()` but first checks whether the warning should in fact be ignored. I am attaching to the PostCSS `result` a prop `disabledRanges` that contains objects with info about the ranges. (Everybody ok with that idea, adding to the `result` object?)

I've been using `block-no-empty` as a prototype rule.

First of all, if anybody has any better naming idea for the functions/objects/etc. added, please share.

I guess I really should write 
- unit tests for the `disableRanges()` function (make sure the `result` has the right `disabledRanges` value, given some CSS with ranges
- unit tests for `report` (make sure that it correctly reads the `disabledRanges` and ignores what should be ignored)
- some more integration tests, e.g. with multiple rules

What do you think so far?

### CODE REVIEW

**Comment by davidtheclark** [comment]
Added those unit tests.

**Comment by davidtheclark** [comment]
Actually now I'm thinking this might be ready for a merge. What would have to happen, though, before it's "done" is that all the rules would have to use `report()` instead of `result.warn()`.

File: src/utils/report.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
+/**
+ * Report a violation.
+ *
+ * @param {object} v - Violation details object
+ * @param {string} v.ruleName - The name of the rule
+ * @param {Result} v.result - PostCSS Result object
+ * @param {string} v.message - Message to inform user of the violation
+ * @param {Node} v.node - PostCSS Node object
+ */
+export default function (v) {
```
Comment (MoOx) [inline src/utils/report.js]: not a fan of using a single letter object but that's a detail ^^

**Comment by MoOx** [comment]
looks really great

File: src/utils/report.js (hunk 1 above)
Comment (davidtheclark) [inline src/utils/report.js]: Just weariness of typing ... I should not succumb :) Happy to rename it "violation" or something.

**Comment by jeddy3** [comment]
This looks ace!

> Actually now I'm thinking this might be ready for a merge. What would have to happen, though, before it's "done" is that all the rules would have to use report() instead of result.warn().

I can help with this, if you like?

**Comment by davidtheclark** [comment]
@jeddy3 if you could help with that refactor, it would be awesome.

**Comment by jeddy3** [comment]
> @jeddy3 if you could help with that refactor, it would be awesome.

Will do. I'm off to bed in a moment, but I can do this tomorrow morning. I'll start from the bottom of the list of rules... less chance of us duplicating any work then :)

I intend to just follow your lead from `block-no-empty`, does that sound about right?

**Comment by davidtheclark** [comment]
Yep. As long as the same tests are passing, I think it should be good.

**Comment by jeddy3** [comment]
@davidtheclark I'm finished for the day. 

I hope you don't mind, but I skipped over the rules that attach the warning to the root like `no-eol-whitespace` as the tests didn't pass first time. If you have a moment can you look into one of those rules for me, and then I'll follow your lead with the other similar rules? Thanks!

**Comment by davidtheclark** [comment]
I had to adjust the report function a little so that these rules could specify a line number (instead of relying on the node). Then the tests passed fine.

**Comment by jeddy3** [comment]
@davidtheclark Thanks for adjusting the report function.

I've refactored the remaining rules to use `report()`. Shall we merge this in?

**Comment by davidtheclark** [comment]
Great!

**Pull request merged**